### DISCUSSION TITLE
Support of more file formats for image data acquisition

### DISCUSSION THREAD

**Discussion by samcondon4** [pymeasure/pymeasure#555] [Ideas]
Currently, there is some support for working with image data (eg. PR #157 ), though the only file format supported for data files generated by pymeasure experiments are csvs. This file format is much less than ideal for image data as one needs to create a formatted line with every single x, y pixel coordinate along with the pixel data. This is fine for small image arrays, as is generated by the default parameters in the `image_gui.py` example that I've linked at the bottom, though it is completely intractable for any realistic sized images (a 1024x768 pixel image takes several minutes for `image_gui.py` to generate, save, and display). 

I believe that more file formats should be supported by pymeasure especially if we want users to be able to work with and record real image data. Here is my current basic understanding of how classes within `pymeasure.experiment` are working for data acquisition and storage (please correct me if any of the following is wrong!):

- `Procedure()`: Class to organize a thread of experiment execution, calling `self.emit("results", <results dictionary>)` to begin the process of writing a line of data to a csv.
- `Results()`: Class to create the initial data file with a header and labels (or load previously created data files and procedures). This class also uses `CSVFormatter()` to format comma delimited strings to append to the csv.
- `Worker()`: Class that runs the user defined procedure `startup()`, `execute()`, and `shutdown()` methods on a separate thread. Also monkey patches the `emit()` method into the user defined procedure class. `emit()` uses the `Recorder()` class to write data to csvs.
-  `Recorder()`: uses the builtin `logging` package to create a child thread that listens for new data and appends the formatted data strings generated by the `CSVFormatter()` to the csv data file.

I would like to begin working on support for additional file formats for working with images in pymeasure. Does anyone have any initial thoughts on starting points for this? As much as possible, I would like to stick with the current way of doing things using (modified versions of) the `Results()`, `Worker()`, and `Recorder()` classes to define data files, execute data logging threads, and log data using python's builtin `logging` package. Though I am unsure if `logging` can be used conveniently with image data (multidimensional numpy arrays). I have found this project: https://github.com/dchaplinsky/visual-logging, for creating html logs with embedded images but am wary of bringing additional dependencies into pymeasure nor do I think that a single log file is the best way to store and work with image data in an experiment.

Has anyone else thought about this? I am also curious to hear the thoughts @jmittelstaedt, who implemented PR #157 

Thanks everyone in advance!   

`image_gui.py` example:
https://github.com/pymeasure/pymeasure/blob/d367e17e031b95888532f9085b73bb48f706ebd3/examples/Basic/image_gui.py

**Comment by LongnoseRob** [comment]
@samcondon4 
I like the idea - not only for image related data -  but also for other arbitrary file formats.
Some example I came up with quickly:
1. S-parameter fiiles (Aka Touchstone standard) from Network-Analyzers
2. modern Oscilloscopes offer high resolution trace data with easyily going to the MB if not GB range
3. IQ sample data from Spektrumanalzyers/Oscilloscopes.

For these three examples the data might could be possibly be transferred with some efforts line-by-line, but it would make no sense.

One possiblity would be to have the procedure to emit a token (e.g. timestamp, instrument-ID, filename) and handle the data transfer outside the procedure, if no real-time analsysis of the data collected is required.

Thank you for opening this discussion and lets see what others opion and ideas are shared..

**Comment by samcondon4** [comment]
I see now that this is a duplicate with #136. Apologies... Not sure if a discussion can be closed like an issue?

**Reply by bilderbuchi** [reply]
Don't worry about it.
